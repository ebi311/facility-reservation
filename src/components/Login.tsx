import Button from '@material-ui/core/Button';
import React, { useCallback } from 'react';
import styled from 'styled-components';
import { auth } from '../auth';

const Container = styled.div`
  width: 100vw;
  height: 100vh;
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const Login: React.FC = () => {
  const onGoogleAuthClick = useCallback(() => {
    auth().then(() => window.location.reload());
  }, []);
  return (
    <Container>
      <Button variant="outlined" onClick={onGoogleAuthClick}>
        Google でログインする
      </Button>
    </Container>
  );
};
