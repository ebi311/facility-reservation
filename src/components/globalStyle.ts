import { createGlobalStyle } from 'styled-components';

const GlobalStyle = createGlobalStyle`
  body, html {
    background-color: #eee;
  }
`;

export default GlobalStyle;
