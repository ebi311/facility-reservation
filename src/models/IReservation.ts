import { Dayjs } from 'dayjs';
import { ISystem } from './ISystem';

export interface IReservation {
  id: string;
  facilityId: string;
  subject: string;
  description: string;
  startDate: Dayjs;
  endDate: Dayjs;
  system: ISystem;
}
