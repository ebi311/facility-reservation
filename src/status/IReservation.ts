import ISystem from './ISystem';
import { Dayjs } from 'dayjs';

export default interface IReservation {
  id: string;
  subject: string;
  facilityId: string;
  description?: string;
  startDate: Dayjs;
  endDate: Dayjs;
  system: ISystem;
}
