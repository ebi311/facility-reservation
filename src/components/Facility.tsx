import Avatar from '@material-ui/core/Avatar/Avatar';
import Button from '@material-ui/core/Button/Button';
import Chip from '@material-ui/core/Chip/Chip';
import Container from '@material-ui/core/Container/Container';
import Grid from '@material-ui/core/Grid/Grid';
import InputLabel from '@material-ui/core/InputLabel/InputLabel';
import Paper from '@material-ui/core/Paper/Paper';
import makeStyles from '@material-ui/core/styles/makeStyles';
import TextField from '@material-ui/core/TextField/TextField';
import DeleteIcon from '@material-ui/icons/Delete';
import DoneIcon from '@material-ui/icons/Done';
import dayjs from 'dayjs';
import React, { useCallback, useEffect, useState } from 'react';
import { Controller, useForm } from 'react-hook-form';
import { useHistory, useParams } from 'react-router-dom';
import { getCurrentUser } from '../auth';
import {
  getFacility,
  postFacility,
  putFacility,
} from '../controllers/facilityController';
import { IFacility } from '../models/IFacility';

const initFacility = (): IFacility => {
  const user = getCurrentUser();
  return {
    id: '',
    name: '',
    note: '',
    system: {
      createDate: new Date(),
      createUser: {
        displayName: user?.displayName || '',
        email: user?.email || '',
        face: user?.photoURL || '',
      },
      lastUpdateUser: {
        displayName: user?.displayName || '',
        email: user?.email || '',
        face: user?.photoURL || '',
      },
      lastUpdate: new Date(),
    },
  };
};

const useStyle = makeStyles((theme) => ({
  root: {
    '& .MuiTextField-root': {
      margin: theme.spacing(1),
    },
  },
  paper: {
    padding: theme.spacing(1),
  },
  rightActions: {
    textAlign: 'right',
  },
  cancelButton: {
    color: theme.palette.error.main,
  },
}));

export const Facility: React.FC = () => {
  const style = useStyle();
  const { id } = useParams<{ id: string }>();
  const [facility, setFacility] = useState(initFacility());
  const { system } = facility;
  const { errors, control, reset, trigger, getValues } = useForm({
    defaultValues: facility,
    mode: 'onBlur',
  });
  useEffect(() => {
    if (!id) return;
    getFacility(id).then((result) => {
      setFacility(result);
      reset(result);
    });
  }, [id]);

  const history = useHistory();
  const onSave = useCallback(async () => {
    const result = await trigger();
    if (!result) return;
    const inputValue = { ...facility, ...getValues() };
    if (!id) {
      // 新規作成
      const id = await postFacility(inputValue);
      history.replace('/facility/' + id);
    } else {
      // 更新
      await putFacility(inputValue);
      window.location.reload();
    }
  }, [id, facility, trigger, getValues]);

  return (
    <Container maxWidth="sm" className={style.root}>
      <Paper className={style.paper}>
        <Controller
          control={control}
          name="name"
          rules={{ required: true }}
          as={
            <TextField
              label="設備名"
              fullWidth
              error={!!errors.name}
              helperText={errors.name ? '必須です' : ''}
            />
          }
        />
        <Controller
          control={control}
          name="note"
          as={<TextField label="詳細" fullWidth multiline value="" />}
        />
        <InputLabel shrink>登録者</InputLabel>
        <p>
          <Chip
            label={system.createUser.displayName}
            avatar={<Avatar src={system.createUser.face} />}
          />
          {dayjs(system.createDate).format('YYYY-MM-DD HH:mm')}
        </p>
        <InputLabel shrink>更新者</InputLabel>
        <p>
          <Chip
            label={system.lastUpdateUser.displayName}
            avatar={<Avatar src={system.lastUpdateUser.face} />}
          />
          {dayjs(system.lastUpdate).format('YYYY-MM-DD HH:mm')}
        </p>
        <Grid container>
          <Grid item xs={6}>
            <Button className={style.cancelButton} startIcon={<DeleteIcon />}>
              削除
            </Button>
          </Grid>
          <Grid item xs={6} className={style.rightActions}>
            <Button
              variant="contained"
              color="primary"
              startIcon={<DoneIcon />}
              onClick={onSave}
            >
              保存
            </Button>
          </Grid>
        </Grid>
      </Paper>
    </Container>
  );
};
