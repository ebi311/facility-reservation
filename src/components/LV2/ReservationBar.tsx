import React from 'react';
import styled from 'styled-components';
import IReservation from '../../status/IReservation';
import moment from 'moment';
import { constraints } from '../../constraints';

type PropsType = {
  reservation: IReservation;
  beginTime: Date;
  timeWidth: number;
  color: string;
};

type BarType = {
  color: string;
  offset: number;
  width: number;
};

const Bar = styled.div<BarType>`
  height: ${constraints.rowHeight / 2}px;
  background-color: ${p => p.color};
  position: absolute;
  left: ${p => p.offset}px;
  top: ${constraints.rowHeight / 4}px;
  width: ${p => p.width}px;
  z-index: 10;
`;

const ReservationBar: React.FC<PropsType> = props => {
  const endDate = moment(props.reservation.endDate);
  const startDate = moment(props.reservation.startDate);
  const startTime = moment(props.beginTime).startOf('hour');
  const offsetTime = startDate.diff(startTime, 'minutes');
  const offsetPixel =
    (offsetTime / 60) * props.timeWidth + constraints.rowHeaderWidth;
  const diffTime = endDate.diff(startDate, 'minutes');
  const width = (diffTime / 60) * props.timeWidth;
  return (
    <>
      <Bar color={props.color} width={width} offset={offsetPixel}></Bar>
    </>
  );
};

export default ReservationBar;
