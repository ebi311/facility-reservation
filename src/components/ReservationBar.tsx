import React, { useCallback, useMemo } from 'react';
import { IReservation } from '../models/IReservation';
import { Property } from 'csstype';
import { makeStyles, Theme } from '@material-ui/core/styles';
import { useHistory } from 'react-router-dom';

type PropsType = {
  reservation: IReservation;
  leftOffset: number;
  beginHour: number;
  hourWidth: number;
  backgroundColor: Property.BackgroundColor;
};

type StyleType = {
  width: number;
  left: number;
  backgroundColor: Property.BackgroundColor;
};

const useStyles = makeStyles<Theme, StyleType>(() => ({
  root: {
    height: '100%',
    position: 'absolute',
    display: 'flex',
    alignItems: 'center',
    width: (p) => p.width + 'px',
    left: (p) => p.left + 'px',
  },
  bar: {
    height: '50%',
    width: '100%',
    backgroundColor: (p) => p.backgroundColor,
    cursor: 'pointer',
  },
}));

export const ReservationBar: React.FC<PropsType> = (props) => {
  const {
    leftOffset,
    reservation,
    hourWidth,
    beginHour,
    backgroundColor,
  } = props;
  const { startDate, endDate } = reservation;

  const width = useMemo(() => {
    const hours = endDate.diff(startDate, 'minute') / 60;
    return hourWidth * hours;
  }, [startDate, endDate, hourWidth]);

  const left = useMemo(() => {
    const beginDate = startDate.set('hour', beginHour).startOf('hour');
    const diffStart = startDate.diff(beginDate, 'minute') / 60;
    return leftOffset + diffStart * hourWidth;
  }, [beginHour, hourWidth, leftOffset, startDate]);

  const style = useStyles({
    width,
    left,
    backgroundColor,
  });

  const history = useHistory();
  const click = useCallback(() => {
    history.push('/reservation/' + reservation.id);
  }, [reservation.id]);

  return (
    <div className={style.root}>
      <div className={style.bar} onClick={click}></div>
    </div>
  );
};
