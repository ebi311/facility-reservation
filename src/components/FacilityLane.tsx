import { makeStyles, Theme } from '@material-ui/core/styles';
import { Property } from 'csstype';
import { Dayjs } from 'dayjs';
import React, { useMemo } from 'react';
import { Link } from 'react-router-dom';
import { IFacility } from '../models/IFacility';
import { IReservation } from '../models/IReservation';
import { ReservationBar } from './ReservationBar';

type Props = JSX.IntrinsicElements['div'] & {
  facility: IFacility;
  cellWidth: number;
  backgroundColor: Property.BackgroundColor;
  reservations: IReservation[];
  date?: Dayjs;
};

const useStyles = makeStyles<
  Theme,
  {
    backgroundColor: Property.BackgroundColor;
  }
>((theme) => ({
  header: {
    backgroundColor: (p) => p.backgroundColor,
    '& a': {
      color: (p) => theme.palette.getContrastText(p.backgroundColor),
    },
  },
}));

export const FacilityLane: React.FC<Props> = (props) => {
  const {
    cellWidth,
    facility,
    reservations,
    backgroundColor,
    date,
    ...rootAttr
  } = props;
  const styles = useStyles({ backgroundColor });
  const cells = useMemo(() => {
    const r: JSX.Element[] = [];
    for (let i = 0; i <= 11; i++) {
      const toDate = date?.hour(i + 8).startOf('hour');
      r.push(
        <Link
          key={i}
          className="timeCell"
          to={`/reservation/?date=${toDate?.toISOString() || ''}`}
        ></Link>,
      );
    }
    return r;
  }, []);
  const bars = useMemo(() => {
    return reservations.map((r) => {
      return (
        <ReservationBar
          key={r.id}
          backgroundColor={backgroundColor}
          beginHour={8}
          reservation={r}
          hourWidth={cellWidth}
          leftOffset={100}
        />
      );
    });
  }, [reservations, backgroundColor, cellWidth]);
  return (
    <div {...rootAttr}>
      {bars}
      <div className={`laneHeader ${styles.header}`}>
        <Link to={'/facility/' + facility.id}>{facility.name}</Link>
      </div>
      {cells}
    </div>
  );
};
