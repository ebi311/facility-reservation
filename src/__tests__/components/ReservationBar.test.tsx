/**
 * @jest-environment jsdom
 */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { render } from '@testing-library/react';
import dayjs from 'dayjs';
import React from 'react';
import { ReservationBar } from '../../components/ReservationBar';
import { IReservation } from '../../models/IReservation';

let useStyles: jest.Mock;
jest.mock('@material-ui/core/styles', () => {
  useStyles = jest.fn().mockImplementation(() => ({
    root: 'root-classname',
    bar: 'bar-classname',
  }));
  return {
    // eslint-disable-next-line @typescript-eslint/ban-types
    ...(jest.requireActual('@material-ui/core/styles') as Object),
    makeStyles: () => useStyles,
  };
});

const reservation: IReservation = {
  description: '',
  endDate: dayjs('2021-07-01T10:00:00+0900'),
  startDate: dayjs('2021-07-01T09:00:00+0900'),
  facilityId: 'f001',
  id: 'r001',
  subject: '予約1',
  system: {} as any,
};

test('ReservationBar', () => {
  const { asFragment } = render(
    <ReservationBar
      backgroundColor="red"
      reservation={reservation}
      beginHour={8}
      hourWidth={100}
      leftOffset={300}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
  expect(useStyles).toBeCalledWith({
    width: 100,
    left: 300 + 100,
    backgroundColor: 'red',
  });
});
