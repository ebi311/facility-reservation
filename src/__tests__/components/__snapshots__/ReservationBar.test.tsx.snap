// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReservationBar 1`] = `
<DocumentFragment>
  <div
    class="root-classname"
  >
    <div
      class="bar-classname"
    />
  </div>
</DocumentFragment>
`;
